routers/api/packages/chef/chef.go

```diff
@@ -0,0 +1,404 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package chef
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/models/db"
+	packages_model "code.gitea.io/gitea/models/packages"
+	"code.gitea.io/gitea/modules/context"
+	packages_module "code.gitea.io/gitea/modules/packages"
+	chef_module "code.gitea.io/gitea/modules/packages/chef"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+	"code.gitea.io/gitea/routers/api/packages/helper"
+	packages_service "code.gitea.io/gitea/services/packages"
+)
+
+func apiError(ctx *context.Context, status int, obj interface{}) {
+	type Error struct {
+		ErrorMessages []string `json:"error_messages"`
+	}
+
+	helper.LogAndProcessError(ctx, status, obj, func(message string) {
+		ctx.JSON(status, Error{
+			ErrorMessages: []string{message},
+		})
+	})
+}
+
+func PackagesUniverse(ctx *context.Context) {
+	pvs, _, err := packages_model.SearchVersions(ctx, &packages_model.PackageSearchOptions{
```

Comment:
Are the results of this search and the below package descriptors search guaranteed to be small?

I guessing that this is common for the packages code?

Reply:
I think your point are possible performance problems? Then yes, this may be a problem for big instances. A cache for the result could be the solution then. For the moment #22491 or #22294 will lower the costs. Compared with some common git operations like directory listings it should still be cheap.

> I guessing that this is common for the packages code?

Not all but many package types offer an endpoint which lists all or a subset of the available packages.

Reply:
My worry is that this may result in a large amount of data being loaded over and over again - (in fact perhaps being doubled in size because of the restructuring in each of these.)

If this is not the only instance of this problem then we can merge this and fix them later though.